File: app/src/main/java/configgen/editorserver/SchemaWriteService.java
package configgen.editorserver;

import configgen.schema.CfgSchema;
import configgen.schema.CfgSchemaErrs;
import configgen.schema.CfgSchemaException;
import configgen.schema.CfgSchemas;
import configgen.schema.cfg.CfgReader;
import configgen.schema.cfg.CfgSyntaxException;
import configgen.util.CachedFiles;
import configgen.util.Logger;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * Schema 文本读写服务：支持获取当前 CFG 文本和写回编辑后的文本。
 */
public class SchemaWriteService {

    public record SchemaText(String text) {
    }

    public record SchemaWriteResult(boolean ok, List<String> errors) {
    }

    /**
     * 读取数据目录下所有 .cfg 文件文本并拼接返回。
     * 根目录的 config.cfg 在最前，各模块 cfg 按包名排序追加。
     */
    public static SchemaText readSchemaText(Path dataDir, List<configgen.schema.CfgFileInfo> cfgFiles) {
        StringBuilder sb = new StringBuilder();
        for (configgen.schema.CfgFileInfo c : cfgFiles) {
            try {
                String content = Files.readString(c.path(), StandardCharsets.UTF_8);
                if (!content.isEmpty()) {
                    if (!sb.isEmpty() && !content.startsWith("\n")) {
                        sb.append("\n");
                    }
                    sb.append(content);
                }
            } catch (IOException e) {
                Logger.log("Failed to read cfg file: " + c.path() + " - " + e.getMessage());
            }
        }
        return new SchemaText(sb.toString());
    }

    /**
     * 解析 CFG 文本，校验语法和 schema 语义，通过后写回根目录的 config.cfg。
     * 多模块场景下，全部内容写入根目录 config.cfg（简化处理，后续可按模块拆分）。
     *
     * @return 写入成功返回 ok=true，失败返回 ok=false + 错误信息列表
     */
    public static SchemaWriteResult writeSchemaText(Path dataDir, String cfgText) {
        List<String> errors = new ArrayList<>();

        // 1. 解析 CFG 文本（语法校验）
        CfgSchema schema;
        try {
            schema = CfgReader.parse(cfgText);
        } catch (CfgSyntaxException e) {
            errors.add(e.getMessage());
            return new SchemaWriteResult(false, errors);
        } catch (Exception e) {
            errors.add(e.getMessage());
            return new SchemaWriteResult(false, errors);
        }

        // 2. Schema 语义校验（resolve 阶段会检查重名、类型引用等）
        try {
            CfgSchemaErrs errs = schema.resolve();
            errs.checkErrors("schemaWrite");
        } catch (CfgSchemaException e) {
            for (CfgSchemaErrs.Err err : e.getErrs().errs()) {
                errors.add(err.msg());
            }
            return new SchemaWriteResult(false, errors);
        } catch (Exception e) {
            errors.add(e.getMessage());
            return new SchemaWriteResult(false, errors);
        }

        // 3. 写回 config.cfg 文件
        Path cfgPath = dataDir.resolve("config.cfg");
        try {
            CachedFiles.writeFile(cfgPath, cfgText.getBytes(StandardCharsets.UTF_8));
            Logger.log("Schema text written to " + cfgPath);
        } catch (IOException e) {
            errors.add("Failed to write config.cfg: " + e.getMessage());
            return new SchemaWriteResult(false, errors);
        }

        return new SchemaWriteResult(true, List.of());
    }
}

File: packages/editor-core/src/SchemaWriteService.ts
/**
 * SchemaWriteService — TypeScript port of Java `configgen.editorserver.SchemaWriteService`.
 *
 * Reads and writes CFG schema text:
 *   - readSchemaText: concatenates all .cfg file contents from the data directory.
 *   - writeSchemaText: parses + validates + writes config.cfg, then reloads the
 *     EditorService context so subsequent reads reflect the new schema.
 *
 * Key differences from Java:
 * - Java methods are static taking (Path dataDir, ...); TS methods take an
 *   EditorService instance and use its context()/rootDir() accessors.
 * - Java's CfgSyntaxException → TS ParseError (from CfgParser).
 * - Java's CfgSchemaException → TS CfgSchemaException (from CfgSchemaErrs.checkErrors()).
 * - Java's writeSchemaText does NOT reload the context — it only parses, validates,
 *   and writes. The caller (EditorServer) is responsible for reloading.
 *   TS follows the same pattern: caller calls editor.reload() if needed.
 * - readSchemaText reads from disk (like Java's Files.readString), not from
 *   context-cached CfgFileInfo.content, so it always reflects the latest file state.
 * - TS writes via fs.writeFileSync (synchronous, matching Java's CachedFiles.writeFile).
 *
 * Java source: configgen.editorserver.SchemaWriteService.java (99 lines)
 */

import * as fs from 'fs';
import * as path from 'path';

import { CfgReader, ParseError, CfgSchemaException, CfgSchemaErrs } from '@cfgforge/schema';
import { getDefaultFileSystem } from '@cfgforge/shared';
import type { EditorService } from './EditorService.js';

// Re-import fs for readSchemaText (reads from disk, not cached content)

// ---------------------------------------------------------------------------
// Types (mirror cfgeditor API expectations)
// ---------------------------------------------------------------------------

export interface SchemaTextResult {
  text: string;
}

export interface SchemaWriteResult {
  ok: boolean;
  errors: string[];
}

// ---------------------------------------------------------------------------
// SchemaWriteService
// ---------------------------------------------------------------------------

export class SchemaWriteService {
  /**
   * Read the concatenated CFG text from all .cfg files in the data directory.
   * Files are sorted by pkgNameDot; root config.cfg comes first.
   *
   * Java: SchemaWriteService.readSchemaText(dataDir, cfgFiles)
   */
  static readSchemaText(editor: EditorService): SchemaTextResult {
    const cfgFiles = editor.context().sourceStructure().getCfgFiles();
    let sb = '';

    for (const c of cfgFiles) {
      try {
        // Read from disk (like Java's Files.readString), not from cached content,
        // so we always reflect the latest file state after a write.
        const content = fs.readFileSync(c.path, 'utf-8');
        if (content.length > 0) {
          if (sb.length > 0 && !content.startsWith('\n')) {
            sb += '\n';
          }
          sb += content;
        }
      } catch {
        // Failed to read — skip (Java logs and continues)
      }
    }

    return { text: sb };
  }

  /**
   * Async variant of readSchemaText.
   * Uses CfgFileSystem abstraction (Tauri/WebView compatible).
   */
  static async readSchemaTextAsync(editor: EditorService): Promise<SchemaTextResult> {
    const cfgFiles = editor.context().sourceStructure().getCfgFiles();
    let sb = '';
    const dfs = getDefaultFileSystem();

    for (const c of cfgFiles) {
      try {
        const bytes = await dfs.readFile(c.path);
        const content = Buffer.from(bytes).toString('utf-8');
        if (content.length > 0) {
          if (sb.length > 0 && !content.startsWith('\n')) {
            sb += '\n';
          }
          sb += content;
        }
      } catch {
        // Failed to read — skip
      }
    }

    return { text: sb };
  }

  /**
   * Parse, validate, and write CFG schema text to config.cfg.
   * On success, reloads the EditorService context to pick up the new schema.
   *
   * Java: SchemaWriteService.writeSchemaText(dataDir, cfgText)
   *
   * @param editor  the EditorService (provides rootDir + reload)
   * @param cfgText the full CFG schema text to write
   * @returns { ok: true } on success, { ok: false, errors: [...] } on failure
   */
  static writeSchemaText(editor: EditorService, cfgText: string): SchemaWriteResult {
    const errors: string[] = [];

    // 1. Parse CFG text (syntax check)
    let schema;
    try {
      schema = CfgReader.parse(cfgText);
    } catch (e) {
      if (e instanceof ParseError) {
        errors.push(e.message);
      } else {
        errors.push((e as Error).message);
      }
      return { ok: false, errors };
    }

    // 2. Schema semantic validation (resolve checks for duplicate names, type refs, etc.)
    try {
      const errs: CfgSchemaErrs = schema.resolve();
      errs.checkErrors('schemaWrite');
    } catch (e) {
      if (e instanceof CfgSchemaException) {
        for (const err of e.errs.errs) {
          errors.push(err.msg());
        }
      } else {
        errors.push((e as Error).message);
      }
      return { ok: false, errors };
    }

    // 3. Write to config.cfg
    const cfgPath = path.join(editor.rootDir(), 'config.cfg');
    try {
      fs.writeFileSync(cfgPath, cfgText, 'utf8');
    } catch (e) {
      errors.push(`Failed to write config.cfg: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    // Note: Java does NOT reload context here — caller is responsible for
    // calling editor.reload() if it needs the context updated.
    return { ok: true, errors: [] };
  }

  /**
   * Async variant of writeSchemaText.
   * Uses CfgFileSystem abstraction (Tauri/WebView compatible).
   */
  static async writeSchemaTextAsync(editor: EditorService, cfgText: string): Promise<SchemaWriteResult> {
    const errors: string[] = [];

    // 1. Parse CFG text (syntax check)
    let schema;
    try {
      schema = CfgReader.parse(cfgText);
    } catch (e) {
      if (e instanceof ParseError) {
        errors.push(e.message);
      } else {
        errors.push((e as Error).message);
      }
      return { ok: false, errors };
    }

    // 2. Schema semantic validation
    try {
      const errs: CfgSchemaErrs = schema.resolve();
      errs.checkErrors('schemaWrite');
    } catch (e) {
      if (e instanceof CfgSchemaException) {
        for (const err of e.errs.errs) {
          errors.push(err.msg());
        }
      } else {
        errors.push((e as Error).message);
      }
      return { ok: false, errors };
    }

    // 3. Write to config.cfg
    const cfgPath = path.join(editor.rootDir(), 'config.cfg');
    try {
      await getDefaultFileSystem().writeFile(cfgPath, Buffer.from(cfgText, 'utf8'));
    } catch (e) {
      errors.push(`Failed to write config.cfg: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    return { ok: true, errors: [] };
  }
}
